rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,700 @@
+
+In other cases, copies can be avoided for the values, but some computation is
+needed:
+
+*   `ListArray<ListArray<...>>` -> `tf.SparseTensor`
+    *   Need to compute the sparse indices from `ListArray`'s list offsets.
+
+The remaining cases require a copy:
+
+*   `ListArray<ListArray<...>>`(of non-equal-length lists) -> dense Tensors
+
+With TensorRepresentation available in the Schema, a TFXIO implementation may
+optimize its decoder to choose the most efficient Arrow type.
+
+#### Conversion of string features
+
+Arrow’s string arrays (`BinaryArray`) have a different memory layout than
+TensorFlow’s string Tensors, even with
+[`tensorflow::tstring`](https://github.com/tensorflow/community/blob/master/rfcs/20190411-string-unification.md).
+There is always some overhead in conversion, but with `tensorflow::tstring` a
+Tensor of `string_view`s is possible, thus the overhead will be a function of
+the number of strings being converted, instead of the lengths of the strings.
+
+#### TF APIs for conversions
+
+In TF 1.x we will use np.ndarray as a bridge as Arrow has zero-copy conversion
+to numpy’s ndarrays. (not for string arrays).
+
+Starting from TF 2.x, we will be able to create EagerTensors from Python
+memoryview(s) so that strings can be covered.
+
+## TFMD Schema
+
+[The TFMD Schema](https://github.com/tensorflow/metadata/blob/master/tensorflow_metadata/proto/v0/schema.proto)
+is a pipeline-level artifact and in the scope of this proposal, it may serve two
+purposes:
+
+*   To provide optional inputs to the parsing logic for optimizations.
+*   To carry user’s intents of converting data to TF feedables.
+
+The two purposes don’t have to be served in the following cases:
+
+*   TFDV should not require a schema to work and it does not need TF feedables.
+*   Some TFXIO implementation may not need the schema for either purposes.
+
+Therefore the TFMD schema is optional, and a TFXIO implementation:
+
+*   should guarantee that the `BeamSource()`can return a valid
+    `PCollection[RecordBatch]` without a schema.
+    *   Other interfaces may raise an error when a schema was not provided.
+*   does not have to require a TFMD schema for all its interfaces to work.
+
+## (TensorFlow) Trainer integration
+
+For TFX to freely choose the storage format for training examples for a user, we
+**cannot** expose file-based or record-based interface to that user in the TF
+trainer, because:
+
+*   the user might not know how to open those files.
+*   there might not be an efficient representation of a “record” (this is true
+    for columnar storage formats like Apache Parquet) but only an efficient
+    representation of a batch of records.
+
+Thus we propose that to most users, the TF Trainer only exposes a handle to a
+`tf.data.Dataset` of parsed (composite) Tensors.
+
+Each `TFXIO` implementation will implement a `TFDataset()` interface to return
+such a `tf.data.Dataset`. This dataset contains logically a set of batched
```

Comment:
Can we instead return a `Dataset` of records, and let the user control their own batching using `Dataset.batch()`? For efficiency, tf.data could easily provide rewrites that push the batching (as well as projections!) down into the reader and map vectorization could push the batching closer to the source.

/cc @jsimsa @rachellim

Reply:
In that case, we need something to represent a record. It could be a one-row RecordBatch. But if rewriting does not happen, that is going to be inefficient..